File: app/4.rb
require 'ostruct'

input = File.read(File.expand_path './inputs/4.txt')

entries = input.split(/\n\n/)

def parse(entry)
  pairs = entry.split(/\s+/)
  hash = Hash[pairs.map { |s| s.split(':') }]
  OpenStruct.new(hash)
end

def height_valid?(hgt)
  match = hgt.match(/(\d+)in|(\d+)cm/)
  return false unless match

  if match[1]
    (59..76).cover?(match[1].to_i)
  elsif match[2]
    (150..193).cover?(match[2].to_i)
  end
end

def valid?(e)
  all_present = %w[byr iyr eyr hgt hcl ecl pid].all? { |k| e[k] }
  return false unless all_present

  byr = (1920..2002).cover?(e.byr.to_i)
  iyr = (2010..2020).cover?(e.iyr.to_i)
  eyr = (2020..2030).cover?(e.eyr.to_i)
  hgt = height_valid?(e.hgt)
  hcl = e.hcl.match(/^#[0-9a-f]{6}$/)
  ecl = %w[amb blu brn gry grn hzl oth].include?(e.ecl)
  pid = e.pid.match(/^\d{9}$/)

  [
    byr,
    iyr,
    eyr,
    hgt,
    hcl,
    ecl,
    pid
  ].all?
end

parsed = entries.map { |e| parse(e) }
p(parsed.count { |e| valid?(e) })

File: app/4.hs
module Main where

import Data.Char
import Data.Map (Map)
import qualified Data.Map as M
import Data.List.Split (splitOn)
import Control.Monad (replicateM)
import Control.Applicative ((<|>))
import Data.Attoparsec.Text
import Data.Text (Text)
import qualified Data.Text as T

type Entry = Map Text Text

pair :: Parser (Text, Text)
pair = do
  key <- T.pack <$> replicateM 3 letter
  char ':'
  val <- T.pack <$> many1 (satisfy (not . isSpace))
  pure (key, val)

entry :: Parser Entry
entry = do
  ps <- pair `sepBy` (char ' ' <|> char '\n')
  pure (M.fromList ps)

requiredKeys =
  [ "byr"
  , "iyr"
  , "eyr"
  , "hgt"
  , "hcl"
  , "ecl"
  , "pid"
  ]

main :: IO ()
main = do
  input <- map T.pack . splitOn "\n\n" <$> readFile "inputs/4.txt"
  let parsed = map (parseOnly entry) input
      parsed' = map (\(Right x) -> x) parsed

  -- Part 1
  print . length . filter (\e -> all (`M.member` e) requiredKeys) $ parsed'
